atomic_reactor/plugins/post_koji_upload.py

```diff
@@ -0,0 +1,636 @@
+
+        if not totalsize or not t1:
+            return
+
+        percent_done = 100 * offset / totalsize
+        if (percent_done >= 99 or
+                percent_done - self.last_percent_done >= self.notable_percent):
+            self.last_percent_done = percent_done
+            self.logger.debug("upload: %d%% done (%.1f MiB/sec)",
+                              percent_done, size / t1 / 1024 / 1024)
+
+
+class KojiUploadPlugin(PostBuildPlugin):
+    """
+    Upload this build to Koji
+
+    Note: only the image archive is uploaded to Koji at this stage.
+    Metadata about this image is created and stored in a ConfigMap in
+    OpenShift, ready for the orchestrator build to collect and use to
+    actually create the Koji Build together with the uploaded image
+    archive(s).
+
+    Authentication is with Kerberos unless the koji_ssl_certs_dir
+    configuration parameter is given, in which case it should be a
+    path at which 'cert', 'ca', and 'serverca' are the certificates
+    for SSL authentication.
+
+    If Kerberos is used for authentication, the default principal will
+    be used (from the kernel keyring) unless both koji_keytab and
+    koji_principal are specified. The koji_keytab parameter is a
+    keytab name like 'type:name', and so can be used to specify a key
+    in a Kubernetes secret by specifying 'FILE:/path/to/key'.
+    """
+
+    key = PLUGIN_KOJI_UPLOAD_PLUGIN_KEY
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow, kojihub, url,
+                 verify_ssl=True, use_auth=True,
+                 koji_ssl_certs_dir=None, koji_proxy_user=None,
+                 koji_principal=None, koji_keytab=None,
+                 blocksize=None):
+        """
+        constructor
+
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param kojihub: string, koji hub (xmlrpc)
+        :param url: string, URL for OSv3 instance
+        :param verify_ssl: bool, verify OSv3 SSL certificate?
+        :param use_auth: bool, initiate authentication with OSv3?
+        :param koji_ssl_certs_dir: str, path to 'cert', 'ca', 'serverca'
+        :param koji_proxy_user: str, user to log in as (requires hub config)
+        :param koji_principal: str, Kerberos principal (must specify keytab)
+        :param koji_keytab: str, keytab name (must specify principal)
+        :param blocksize: int, blocksize to use for uploading files
+        """
+        super(KojiUploadPlugin, self).__init__(tasker, workflow)
+
+        self.kojihub = kojihub
+        self.koji_ssl_certs_dir = koji_ssl_certs_dir
+        self.koji_proxy_user = koji_proxy_user
+
+        self.koji_principal = koji_principal
+        self.koji_keytab = koji_keytab
+
+        self.blocksize = blocksize
+
+        self.namespace = get_build_json().get('metadata', {}).get('namespace', None)
+        osbs_conf = Configuration(conf_file=None, openshift_uri=url,
```

Comment:
We also need to set the `build_json_dir` parameter here.
This is because osbs library will load the `config_map.json` as part of creating the ConfigMap object in OpenShift.

See how `import_image` plugin does this: https://github.com/projectatomic/atomic-reactor/blob/master/atomic_reactor/plugins/post_import_image.py#L77

In osbs-client you also want to make sure this is set correctly. 
See how `import_image` plugin is rendered in `render_import_image` method of `build_request.py`.
(Not sure if you or @mlangsdorf is doing the osbs-client portion)